edb/server/conn_pool/src/python.rs

```diff
@@ -0,0 +1,371 @@
+use crate::{
+    algo::PoolConstraints,
+    conn::{ConnError, Connector},
+    pool::{Pool, PoolConfig},
+    PoolHandle,
+};
+use futures::TryFutureExt;
+use pyo3::{
+    exceptions::PyException,
+    prelude::*,
+    types::{PyDict, PyTuple},
+};
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    rc::Rc,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc, RwLock,
+    },
+};
+use tokio::task::LocalSet;
+use tracing::{error, trace};
+use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt};
+
+pyo3::create_exception!(_conn_pool, InternalError, PyException);
+
+#[derive(Debug)]
+#[repr(u8)]
+enum ConnectOp {
+    Connect,
+    Disconnect,
+    Reconnect,
+}
+
+#[derive(Debug, Default)]
+struct PythonConnectionMap {
+    /// Connection : [`PoolHandle`] (to keep the handle alive)
+    handle: HashMap<usize, PoolHandle<PythonConnectionFactory>>,
+    py_dict: Option<Py<PyDict>>,
+    next_id: usize,
+}
+
+impl PythonConnectionMap {
+    pub fn insert(&mut self, py: Python, handle: PoolHandle<PythonConnectionFactory>) {
+        let py_dict = self
+            .py_dict
+            .get_or_insert_with(|| PyDict::new(py).into())
+            .as_ref(py);
+        _ = handle.with_handle(|conn| py_dict.set_item(conn, self.next_id));
+        self.handle.insert(self.next_id, handle);
+        self.next_id += 1;
+    }
+
+    pub fn remove(
+        &mut self,
+        py: Python,
+        conn: PyObject,
+    ) -> Option<PoolHandle<PythonConnectionFactory>> {
+        let Some(py_dict) = &mut self.py_dict else {
+            return None;
+        };
+        let py_dict = py_dict.as_ref(py);
+        let item = py_dict.get_item(conn.clone_ref(py)).ok()??;
+        _ = py_dict.del_item(conn);
```

Comment:
Python dicts have also pop_item() / pop() methods, which could be useful here

Reply:
Ah, yeah that could be useful. I may need to PR some bindings for that in PyO3, or extract and save the dict.pop method. I'll check the CPython binding API to see if it is possible to call.

Reply:
It's a minor nit comment though, ultimately it doesn't matter.